### PR TITLE
[SPARK-19059] [SQL] Unable to retrieve data from parquet table whose name startswith underscore

### PR DESCRIPTION
**Pull request by jayadevanmurali**
## What changes were proposed in this pull request?
The initial shouldFilterOut() method invocation filter the root path name(table name in the intial call) and remove if it contains _. I moved the check one level below, so it first list files/directories in the given root path and then apply filter.
(Please fill in changes proposed in this fix)

## How was this patch tested?
Added new test case for this scenario
(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)
(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)

Please review http://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

**Comment by jayadevanmurali** [comment]
@uncleGen @Eric Liang @cloud-fan 

Could you please review the PR

**Review by gatorsmile — COMMENTED** [review]
See the related JIRA: https://issues.apache.org/jira/browse/HIVE-6431

> By default, FileInputFormat(which is the super class of various formats) in hadoop ignores file name starts with "_" or ".", and hard to walk around this in hive codebase.

We are also using `FileInputFormat `. Thus, it means we are facing the same issue.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 1]
```diff
@@ -2497,9 +2497,7 @@ class SQLQuerySuite extends QueryTest with SharedSQLContext {
       } finally {
         newSession.sessionState.conf.setConf(SQLConf.RUN_SQL_ON_FILES, originalValue)
       }
-    }
-  }
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: does is still compile?

**Comment by cloud-fan** [comment]
ok to test

**Comment by jayadevanmurali** [comment]
retest this please

**Comment by jayadevanmurali** [comment]
retest this please

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 2]
```diff
@@ -2513,4 +2513,18 @@ class SQLQuerySuite extends QueryTest with SharedSQLContext {
       }
     }
   }
+
+  test(
+    "SPARK-19059: Unable to retrieve data from parquet table whose name startswith underscore") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: this bug is not parquet only right? how about `SPARK-19059: read file based table whose name starts with underscore`

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 3]
```diff
@@ -2513,4 +2513,18 @@ class SQLQuerySuite extends QueryTest with SharedSQLContext {
...
+    sql("CREATE TABLE `_tbl`(i INT) USING parquet")
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: use `withTable("tbl", "_tbl") {...}`

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 4]
```diff
@@ -2513,4 +2513,18 @@ class SQLQuerySuite extends QueryTest with SharedSQLContext {
...
+    sql("INSERT INTO `_tbl` VALUES (1), (2), (3)")
+    checkAnswer( sql("SELECT * FROM `_tbl`"), Row(1) :: Row(2) :: Row(3) :: Nil)
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: I think we can stop here, create a table with underscore, then insert into it, then read it, that's enough to prove we can support table with underscore.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala (hunk 2 above)
Comment (jayadevanmurali) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: Yes, it is not parquet only. I think SPARK-19059: read file based table whose name starts with underscore is fine.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala (hunk 3 above)
Comment (jayadevanmurali) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: I will do this change

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala (hunk 4 above)
Comment (jayadevanmurali) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: It should be fine.

**Comment by jayadevanmurali** [comment]
@cloud-fan 
Incorporated code review comments

**Comment by jayadevanmurali** [comment]
retest this please

**Comment by uncleGen** [comment]
process was terminated by signal 9

**Comment by uncleGen** [comment]
retest  this please.

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by springcoil** [comment]
Came across this error today in Spark 2.1.0. Just adding this as a remark. I had code like `nz_exp_test = spark.read.parquet("dbfs:/experience_test/_NZ_2017-05-31-11-59-05.parquet")` the error reported wasn't very useful it was ` AnalysisException: unable to infer schema for Parquet. It must be specified manually.;'`

**Pull request closed without merging**